File: backend-nestjs/src/modules/zones/application/commands/update-zone.handler.ts
import { CommandHandler, ICommandHandler } from '@nestjs/cqrs';
import { Inject, NotFoundException } from '@nestjs/common';
import type { Cache } from 'cache-manager';
import { CACHE_MANAGER } from '@nestjs/cache-manager';
import { UpdateZoneCommand } from './update-zone.command';
import { ZoneRepository } from '../../infrastructure/zone.repository';
import { Zone } from '../../domain/zone';

@CommandHandler(UpdateZoneCommand)
export class UpdateZoneHandler implements ICommandHandler<
  UpdateZoneCommand,
  Zone
> {
  constructor(
    private readonly repository: ZoneRepository,
    @Inject(CACHE_MANAGER) private readonly cache: Cache,
  ) {}

  async execute(command: UpdateZoneCommand): Promise<Zone> {
    const zone = await this.repository.findById(command.id);
    if (!zone) throw new NotFoundException(`Zone ${command.id} not found`);

    if (command.name !== undefined) zone.rename(command.name);
    if (command.type !== undefined) zone.updateType(command.type);
    if (command.lat !== undefined || command.lng !== undefined) {
      zone.reposition(
        command.lat ?? zone.coordinates.lat,
        command.lng ?? zone.coordinates.lng,
      );
    }
    if (command.targetCount !== undefined)
      zone.updateTargetCount(command.targetCount);
    if (command.currentCount !== undefined)
      zone.updateProgress(command.currentCount);
    if (command.status !== undefined) zone.changeStatus(command.status);
    if (command.description !== undefined)
      zone.updateDescription(command.description);
    if (command.organizerContact !== undefined)
      zone.updateOrganizerContact(command.organizerContact);
    if (command.treeSpecies !== undefined)
      zone.updateTreeSpecies(command.treeSpecies);

    const saved = await this.repository.save(zone);
    await this.cache.del('zones:all');
    await this.cache.del(`zone:${command.id}`);
    return saved;
  }
}

File: backend-springboot/src/main/java/com/greenalgeria/zone/application/command/UpdateZoneHandler.java
package com.greenalgeria.zone.application.command;

import com.greenalgeria.zone.application.*;
import com.greenalgeria.zone.domain.Coordinates;
import com.greenalgeria.zone.domain.ZoneRepository;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

@Component
@Transactional
public class UpdateZoneHandler {

    private final ZoneRepository zoneRepository;

    public UpdateZoneHandler(ZoneRepository zoneRepository) {
        this.zoneRepository = zoneRepository;
    }

    public ZoneResponse handle(UpdateZoneCommand command) {
        var zone = zoneRepository
                .findById(command.id())
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Zone not found"));
        var request = command.request();
        if (request.name() != null) zone.rename(request.name());
        if (request.type() != null) zone.setType(request.type());
        if (request.status() != null) {
            switch (request.status()) {
                case in_progress -> zone.markInProgress();
                case completed -> zone.markComplete();
                default -> {}
            }
        }
        if (request.lat() != null && request.lng() != null) {
            zone.reposition(new Coordinates(request.lat(), request.lng()));
        }
        if (request.targetCount() != null) zone.setTargetCount(request.targetCount());
        if (request.currentCount() != null) zone.setCurrentCount(request.currentCount());
        if (request.description() != null) zone.setDescription(request.description());
        if (request.organizerContact() != null) zone.setOrganizerContact(request.organizerContact());
        if (request.treeSpecies() != null) zone.setTreeSpecies(request.treeSpecies());
        return ZoneResponse.from(zoneRepository.save(zone));
    }
}
